notebooks/working/ho_deconfounder_on_sim_data.py

```diff
@@ -0,0 +1,512 @@
+
+# +
+## Return the variables of each utility specification
+
+spec_dic = specifications(mnl_specification=mnl_specification, num_modes=8)
+spec_dic
+# -
+
+confounder_vectors = []
+holdout_dfs = []
+masks_df = []
+rows_df =[]
+latent_dim = 1
+for i in data['mode_id'].unique():
+
+    data_mode_i = data[data['mode_id']==i]
+    # standardize the data for PPCA
+    print("Analysis for mode: ", i)
+    print("-------------------------------------------------------------------------------------------")
+    X_columns = spec_dic[i]
+    
+    if 'intercept' in X_columns:
+        X_columns.remove('intercept')
+        
+    if 'total_travel_distance' in X_columns:
+        X_columns.remove('total_travel_distance')
+        
+    print(i, X_columns)
+    
+    X = np.array((data_mode_i[X_columns] - data_mode_i[X_columns].mean())/data_mode_i[X_columns].std())
+    
+    confounders, holdouts, holdoutmasks, holdoutrow= confounder_ppca(holdout_portion=0.2, X=X, latent_dim=latent_dim)
+
+    confounder_vectors.append(confounders)
+    holdout_dfs.append(holdouts)
+    masks_df.append(holdoutmasks)
+    rows_df.append(holdoutrow)
+
+# ## Method 2
+#
+# In this method, we only run the deconfounder's factor model on travel_time and travel_cost, which are the only variables confounded with travel_distance in our simulation
+
+# +
+xs = ['total_travel_time', 'total_travel_cost']
+X_DA = (data[data['mode_id']==1][xs] - data[data['mode_id']==1][xs].mean())/data[data['mode_id']==1][xs].std()
+
+X_SR2 = (data[data['mode_id']==2][xs] - data[data['mode_id']==2][xs].mean())/data[data['mode_id']==2][xs].std()
+
+X_SR3 = (data[data['mode_id']==3][xs] - data[data['mode_id']==3][xs].mean())/data[data['mode_id']==3][xs].std()
+
+(confounders_DA, holdouts_DA, 
+ holdoutmasks_DA, holdoutrow_DA)= confounder_ppca(holdout_portion=0.2, 
+                                       X=X_DA, latent_dim=latent_dim)
+
+(confounders_SR2, holdouts_SR2, 
+ holdoutmasks_SR2, holdoutrow_SR2)= confounder_ppca(holdout_portion=0.2, 
+                                       X=X_SR2, latent_dim=latent_dim)
+
+(confounders_SR3, holdouts_SR3, 
+ holdoutmasks_SR3, holdoutrow_SR3)= confounder_ppca(holdout_portion=0.2, 
+                                       X=X_SR3, latent_dim=latent_dim)
+
+
+
+
+# -
+
+# ## Adding confounders to original DF
+
+# +
+for i in data['mode_id'].unique():
+    
+#     print(len(data.loc[data['mode_id']==i, col_name]), len(confounder_vectors[int(i-1)][2]) )
+    
+    col_name = 'confounder_for_mode_' + str(int(i))
+    data.loc[data['mode_id']==i, col_name] = confounder_vectors[int(i-1)][2]
+    data[col_name] = data[col_name].fillna(0)
+    
+data['confounder_all'] = data[['confounder_for_mode_1','confounder_for_mode_2','confounder_for_mode_3',
+                              'confounder_for_mode_4', 'confounder_for_mode_5', 'confounder_for_mode_6',
+                              'confounder_for_mode_7', 'confounder_for_mode_8']].sum(axis=1)
```

Comment:
I think this block of code should be separated into its own function for method 2 and defined outside the python notebook. The `confounder_all` column should be briefly described in the notebook though. It make sense what the `confounder_all` does and why it'll work when one uses alternative specific coefficients, but isn't immediately clear, especially to someone who isn't familiar with PyLogit.